File: swift-bridge/Sources/ImageIOBridge/Properties.swift
import Foundation
import ImageIO

@_cdecl("imageio_mutable_properties_create")
public func imageioMutablePropertiesCreate() -> UnsafeMutableRawPointer? {
    retainBox(NSMutableDictionary())
}

@_cdecl("imageio_mutable_properties_freeze")
public func imageioMutablePropertiesFreeze(_ raw: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let raw else {
        return nil
    }
    let dictionary = unretainedBox(raw, as: NSMutableDictionary.self).value
    return retainBox(NSDictionary(dictionary: dictionary))
}

@_cdecl("imageio_mutable_properties_set_string")
public func imageioMutablePropertiesSetString(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ value: UnsafePointer<CChar>?
) {
    guard let raw, let key, let value else {
        return
    }
    let dictionary = unretainedBox(raw, as: NSMutableDictionary.self).value
    dictionary[String(cString: key)] = String(cString: value)
}

@_cdecl("imageio_mutable_properties_set_i64")
public func imageioMutablePropertiesSetI64(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ value: Int64
) {
    guard let raw, let key else {
        return
    }
    let dictionary = unretainedBox(raw, as: NSMutableDictionary.self).value
    dictionary[String(cString: key)] = NSNumber(value: value)
}

@_cdecl("imageio_mutable_properties_set_f64")
public func imageioMutablePropertiesSetF64(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ value: Double
) {
    guard let raw, let key else {
        return
    }
    let dictionary = unretainedBox(raw, as: NSMutableDictionary.self).value
    dictionary[String(cString: key)] = NSNumber(value: value)
}

@_cdecl("imageio_mutable_properties_set_bool")
public func imageioMutablePropertiesSetBool(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ value: Bool
) {
    guard let raw, let key else {
        return
    }
    let dictionary = unretainedBox(raw, as: NSMutableDictionary.self).value
    dictionary[String(cString: key)] = NSNumber(value: value)
}

@_cdecl("imageio_mutable_properties_set_dictionary")
public func imageioMutablePropertiesSetDictionary(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ nested: UnsafeMutableRawPointer?
) {
    guard let raw, let key, let nested else {
        return
    }
    let dictionary = unretainedBox(raw, as: NSMutableDictionary.self).value
    let nestedDictionary = unretainedBox(nested, as: NSDictionary.self).value
    dictionary[String(cString: key)] = NSDictionary(dictionary: nestedDictionary)
}

@_cdecl("imageio_properties_copy_keys")
public func imageioPropertiesCopyKeys(_ raw: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let raw else {
        return nil
    }
    let dictionary = unretainedBox(raw, as: NSDictionary.self).value
    let keys = dictionary.allKeys.compactMap { $0 as? String }.sorted()
    return retainBox(keys)
}

@_cdecl("imageio_properties_has_key")
public func imageioPropertiesHasKey(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?
) -> Bool {
    guard let raw, let key else {
        return false
    }
    let dictionary = unretainedBox(raw, as: NSDictionary.self).value
    return dictionary[String(cString: key)] != nil
}

@_cdecl("imageio_properties_copy_string")
public func imageioPropertiesCopyString(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?
) -> UnsafeMutableRawPointer? {
    guard let raw, let key else {
        return nil
    }
    let dictionary = unretainedBox(raw, as: NSDictionary.self).value
    guard let value = dictionary[String(cString: key)] as? String else {
        return nil
    }
    return retainBox(value)
}

@_cdecl("imageio_properties_get_i64")
public func imageioPropertiesGetI64(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ outValue: UnsafeMutablePointer<Int64>?
) -> Bool {
    guard let raw, let key, let outValue else {
        return false
    }
    let dictionary = unretainedBox(raw, as: NSDictionary.self).value
    guard let number = dictionary[String(cString: key)] as? NSNumber else {
        return false
    }
    outValue.pointee = number.int64Value
    return true
}

@_cdecl("imageio_properties_get_f64")
public func imageioPropertiesGetF64(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ outValue: UnsafeMutablePointer<Double>?
) -> Bool {
    guard let raw, let key, let outValue else {
        return false
    }
    let dictionary = unretainedBox(raw, as: NSDictionary.self).value
    guard let number = dictionary[String(cString: key)] as? NSNumber else {
        return false
    }
    outValue.pointee = number.doubleValue
    return true
}

@_cdecl("imageio_properties_get_bool")
public func imageioPropertiesGetBool(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ outValue: UnsafeMutablePointer<Bool>?
) -> Bool {
    guard let raw, let key, let outValue else {
        return false
    }
    let dictionary = unretainedBox(raw, as: NSDictionary.self).value
    guard let number = dictionary[String(cString: key)] as? NSNumber else {
        return false
    }
    outValue.pointee = number.boolValue
    return true
}

@_cdecl("imageio_properties_copy_dictionary")
public func imageioPropertiesCopyDictionary(
    _ raw: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?
) -> UnsafeMutableRawPointer? {
    guard let raw, let key else {
        return nil
    }
    let dictionary = unretainedBox(raw, as: NSDictionary.self).value
    guard let nested = dictionary[String(cString: key)] as? NSDictionary else {
        return nil
    }
    return retainBox(NSDictionary(dictionary: nested))
}

File: src/properties.rs
//! Generic `ImageIO` property dictionaries.

use crate::bridge::{self, properties as ffi, Handle};
use crate::error::ImageError;

/// Immutable `ImageIO` property dictionary.
#[derive(Debug)]
pub struct ImageProperties {
    raw: Handle,
}

impl ImageProperties {
    pub(crate) fn from_raw(raw: Handle) -> Option<Self> {
        (!raw.is_null()).then_some(Self { raw })
    }

    pub(crate) const fn as_raw(&self) -> Handle {
        self.raw
    }

    #[must_use]
    /// Returns the dictionary keys available in this `ImageIO` property map.
    pub fn keys(&self) -> Vec<String> {
        bridge::copy_string_array(unsafe { ffi::imageio_properties_copy_keys(self.raw) })
    }

    /// Checks whether a property key is present.
    pub fn has_key(&self, key: &str) -> Result<bool, ImageError> {
        let key = bridge::cstring(key)?;
        Ok(unsafe { ffi::imageio_properties_has_key(self.raw, key.as_ptr()) })
    }

    /// Reads a string property.
    pub fn string(&self, key: &str) -> Result<Option<String>, ImageError> {
        let key = bridge::cstring(key)?;
        Ok(bridge::copy_string(unsafe {
            ffi::imageio_properties_copy_string(self.raw, key.as_ptr())
        }))
    }

    /// Reads an integer property backed by `CFNumberRef`.
    pub fn i64(&self, key: &str) -> Result<Option<i64>, ImageError> {
        let key = bridge::cstring(key)?;
        let mut value = 0_i64;
        let found = unsafe { ffi::imageio_properties_get_i64(self.raw, key.as_ptr(), &mut value) };
        Ok(found.then_some(value))
    }

    /// Reads a floating-point property backed by `CFNumberRef`.
    pub fn f64(&self, key: &str) -> Result<Option<f64>, ImageError> {
        let key = bridge::cstring(key)?;
        let mut value = 0.0_f64;
        let found = unsafe { ffi::imageio_properties_get_f64(self.raw, key.as_ptr(), &mut value) };
        Ok(found.then_some(value))
    }

    /// Reads a Boolean property backed by `CFBooleanRef`.
    pub fn bool(&self, key: &str) -> Result<Option<bool>, ImageError> {
        let key = bridge::cstring(key)?;
        let mut value = false;
        let found = unsafe { ffi::imageio_properties_get_bool(self.raw, key.as_ptr(), &mut value) };
        Ok(found.then_some(value))
    }

    /// Reads a nested `ImageIO` property dictionary.
    pub fn dictionary(&self, key: &str) -> Result<Option<Self>, ImageError> {
        let key = bridge::cstring(key)?;
        Ok(Self::from_raw(unsafe {
            ffi::imageio_properties_copy_dictionary(self.raw, key.as_ptr())
        }))
    }
}

impl Clone for ImageProperties {
    fn clone(&self) -> Self {
        Self {
            raw: bridge::retain(self.raw),
        }
    }
}

impl Drop for ImageProperties {
    fn drop(&mut self) {
        bridge::release(self.raw);
    }
}

/// Mutable property dictionary builder.
#[derive(Debug)]
pub struct MutableProperties {
    raw: Handle,
}

impl MutableProperties {
    /// Creates a mutable property dictionary for `CGImageDestinationSetProperties`.
    pub fn new() -> Result<Self, ImageError> {
        let raw = unsafe { ffi::imageio_mutable_properties_create() };
        (!raw.is_null()).then_some(Self { raw }).ok_or_else(|| {
            ImageError::Unknown("imageio_mutable_properties_create returned NULL".into())
        })
    }

    /// Sets a string property.
    pub fn set_string(&mut self, key: &str, value: &str) -> Result<(), ImageError> {
        let key = bridge::cstring(key)?;
        let value = bridge::cstring(value)?;
        unsafe {
            ffi::imageio_mutable_properties_set_string(self.raw, key.as_ptr(), value.as_ptr());
        }
        Ok(())
    }

    /// Sets an integer property.
    pub fn set_i64(&mut self, key: &str, value: i64) -> Result<(), ImageError> {
        let key = bridge::cstring(key)?;
        unsafe { ffi::imageio_mutable_properties_set_i64(self.raw, key.as_ptr(), value) };
        Ok(())
    }

    /// Sets a floating-point property.
    pub fn set_f64(&mut self, key: &str, value: f64) -> Result<(), ImageError> {
        let key = bridge::cstring(key)?;
        unsafe { ffi::imageio_mutable_properties_set_f64(self.raw, key.as_ptr(), value) };
        Ok(())
    }

    /// Sets a Boolean property.
    pub fn set_bool(&mut self, key: &str, value: bool) -> Result<(), ImageError> {
        let key = bridge::cstring(key)?;
        unsafe { ffi::imageio_mutable_properties_set_bool(self.raw, key.as_ptr(), value) };
        Ok(())
    }

    /// Sets a nested `ImageIO` property dictionary.
    pub fn set_dictionary(&mut self, key: &str, value: &ImageProperties) -> Result<(), ImageError> {
        let key = bridge::cstring(key)?;
        unsafe {
            ffi::imageio_mutable_properties_set_dictionary(self.raw, key.as_ptr(), value.as_raw());
        }
        Ok(())
    }

    /// Freezes this builder into an immutable property dictionary.
    pub fn freeze(&self) -> Result<ImageProperties, ImageError> {
        ImageProperties::from_raw(unsafe { ffi::imageio_mutable_properties_freeze(self.raw) })
            .ok_or_else(|| {
                ImageError::Unknown("imageio_mutable_properties_freeze returned NULL".into())
            })
    }
}

impl Clone for MutableProperties {
    fn clone(&self) -> Self {
        Self {
            raw: bridge::retain(self.raw),
        }
    }
}

impl Drop for MutableProperties {
    fn drop(&mut self) {
        bridge::release(self.raw);
    }
}
